File: server/NycDispatch.Api/Models/Assignment.cs
namespace NycDispatch.Api.Models;

/// <summary>
/// A mandatory-response staffing level ("assignment") a call is dispatched at.
/// Assignments form an escalation chain via <see cref="UpgradeTo"/>: once a call
/// is fully staffed for its current assignment and has been worked for a while,
/// it may roll <see cref="UpgradeProbability"/> to escalate to the next level.
/// </summary>
public record Assignment
{
    public required string Id { get; init; }
    public required string Name { get; init; }
    /// <summary>Probability (0-1) of escalating to <see cref="UpgradeTo"/> once fully staffed.</summary>
    public required double UpgradeProbability { get; init; }
    /// <summary>Id of the next assignment in the escalation chain, or null if this is the top.</summary>
    public string? UpgradeTo { get; init; }
    /// <summary>Minimum on-scene work time (game-seconds) before the call can auto-resolve.</summary>
    public required int MinResolveS { get; init; }
    /// <summary>Maximum on-scene work time (game-seconds) before the call auto-resolves.</summary>
    public required int MaxResolveS { get; init; }
    public required int Engine { get; init; }
    public required int Ladder { get; init; }
    public required int Rescue { get; init; }
    public required int Squad { get; init; }
    public required int Squad2piece { get; init; }
    public required int Battalion { get; init; }
    public required int Division { get; init; }
    public required int Rac { get; init; }
    public required int Satellite { get; init; }
    public required int Tsu { get; init; }
    public required int Msu { get; init; }
    public required int Fieldcomm { get; init; }
    public required int Mcp { get; init; }
    public required int Hazmat { get; init; }
    public required int Htmu { get; init; }
    public required int Hazmatsupport { get; init; }
    public required int Hazmatbattalion { get; init; }
    public required int Rescuebattalion { get; init; }
    public required int Safetybattalion { get; init; }
    public required int Brush { get; init; }
    public required int Collapse { get; init; }
    public required int Purplek { get; init; }
    public required int Imt { get; init; }
    public required int Highrise { get; init; }
    public required int Thawing { get; init; }
}

File: src/models/Assignment.ts
/** Minimum unit counts, by category, required to be on scene for an assignment. */
export interface AssignmentRequirements {
  engine: number;
  ladder: number;
  rescue: number;
  squad: number;
  squad2piece: number;
  battalion: number;
  division: number;
  rac: number;
  satellite: number;
  tsu: number;
  msu: number;
  fieldcomm: number;
  mcp: number;
  hazmat: number;
  htmu: number;
  hazmatsupport: number;
  hazmatbattalion: number;
  rescuebattalion: number;
  safetybattalion: number;
  brush: number;
  collapse: number;
  purplek: number;
  imt: number;
  highrise: number;
  thawing: number;
}

/**
 * A mandatory-response staffing level (e.g. "Standard", "All Hands"). Calls are
 * dispatched at an initial assignment and may escalate to `upgradeTo` once fully
 * staffed and worked for a while, rolling `upgradeProbability` each check.
 */
export interface Assignment extends AssignmentRequirements {
  id: string;
  name: string;
  upgradeProbability: number;
  upgradeTo: string | null;
  /** Minimum on-scene work time (game-seconds) before the call can auto-resolve. */
  minResolveS: number;
  /** Maximum on-scene work time (game-seconds) before the call auto-resolves. */
  maxResolveS: number;
}
